### ISSUE TITLE
tcp client: EACCESS on resume

### ISSUE THREAD

**Issue by gdt** [MarcusWolschon/osmeditor4android#1254]
## Vespucci Version
git master from the last few days, built locally

## Device (Manufacturer and Model)
Pixel XL

## Android Version 
Lineage 15.1 / AOSP 8.1

## Behaviour/Symptoms
With TCP client configured, and the device present or not present, run Vespucci.  Use the home button.  After a short while (e.g 5s), start Vespucci again.   See a "bind failed EACCESS permission denied" that is too fast to screenshot.

However, I do get position information in this case.

## Expected Behaviour
No error.

## Theory
My guess is that when Vespucci is suspended by back button (really, not being foreground), then it should stop using GPS, if internal, and the same logic applies to TCP client.  So it should close the socket.   Then later when it is restarted it should do connect() again.  I am guessing that the socket is not closed, and there is a second connect.  Or perhaps it is closed, and the socket is in CLOSE_WAIT or something.  But because things work, I am guessing the socket was still connected.

## How to recreate
(Described in Behavior)

## Crash dump submitted (no or yes + date)
no

**Comment by gdt** [comment]
I just ran this again with adb.  All of this is with 1) not recording a track and 2) my RTK device powered off.

If I use the "back" button and click through the "are you sure you want to exit" dialog, I see that TrackerService is torn down and I see "NmeaTcpClient: Cancel called" in the log.  When I start vespucci, it creates a socket that tries to connect and all is ok.

If I use the "home" button, then I see onPause onStop, then "TrackerService: Enabling GPS update" then "NmeaTcpClientServer: Exception  java.net.BindException: bind failed: EACCES (Permission denied)".

So I think that there is an inconsistency about turning off tracker when switching away from vespucci by the home button.  With internal GPS, I think it is stopped if not recording GPX, and kept if recording.  So it seems that the NmeaTcpClient cancel method should be invoked.

I am trying to read the code to figure out what should be changed.

So the repro recipe is:

  configure a tcp client for NMEA

  force-quit vespucci

  with the tcp client not accessible, start vespucci

  wait 20s

  use the home button

  wait 20s

  push the vespucci icon on the home screen

**Comment by simonpoole** [comment]
The home button does "nothing", that is Android will pause the top level activity of the app, and after a while might decide to remove it completely, but there are no guarantees.  Track recording will run the TrackerService as "foreground" service that is persistent, and which, except in dire circumstances, will continue running regardless of what state the rest of the app is in. Both "tcp clients" run as separate threads in that service so shouldn't be reconnecting in that scenario, which is consistent with what you are seeing.

I suspect adding cancelNmeaClients() to stop() before stopSelf() (that is at line 453/454 in TrackerSerive.java) should do the trick, if you could test that that would be good.

**Comment by gdt** [comment]
Thanks for the hints.  I have been reading the code and believe it may be more complicated than that but I will report back, ideally with a tested fix, after I have read enough to really understand.

**Comment by simonpoole** [comment]
Any progress? I can check it out myself, but that means that I need to setup the GPS part of the story.

**Comment by gdt** [comment]
Slight progress - was going to wait until I figured it out.  Am still happily debugging when this is paged in, vs going out mapping if the weather is ok, other projects, etc.  Despite the error I am able to gather tracks and add points with no difficulty.

I have added the close as you suggest and still see the EACCESS.   My current belief is that there is  a plan  about the service staying active and the threads, and that the cancel does not stop the threads, so on restart new threads get spawned because the thread variable has  been set to null.  Overall my thought is that "close the TCP connection" should be like withdrawing active listening from the location API, and not necessarily tear down the threads.  Or at least the thread variable can't be cleared unless the thread has exited.  But I'm not at all sure I have that right, and I was going to document the invariants and assumptions by code reading in one commit, expecting that having done that the fix will be obvious.

**Comment by simonpoole** [comment]
OK, I'll give it another look. Need to jog my memory.

**Comment by simonpoole** [comment]
@gdt could you give https://github.com/MarcusWolschon/osmeditor4android/commit/3baec3fa6a6a592bfff5a26b22bddf02cd9018fb a spin, should fix the issue (and hopefully doesn't break anything).

**Comment by gdt** [comment]
Thanks for working on my bug report.

I tested master from about 4 hours ago, which has the commit you mentioned.    Things are *vastly* improved; I had no vespucci crashes even with switching back and forth, and no loss of positioning.

The only things related to this bug I noticed are:
  * It seems if I back button and confirm, while not tracklogging, the connection stays (because my incoming TCP blinky light on the ardusimple wifi ntrip master is still active).  Have not really figured this out for sure; I could well be confused.  But restarting results in me having position, so no functional issue.
  * If I record some gpx and make a few nodes, stop gpx, export gpx and then go to save josm, I get the file chooser and when I come back to Vespucci I get a "EACCESS", but positioning is working.  So it feels like the plans of whether to close or not TCP while switching  away to file chooser and whether to restart it may not agree, and this is perhaps the last vestige of the issue.

I'll be using it heavily on the weekend, so I can update the ticket after that.

**Comment by gdt** [comment]
For pedantry I should say that I tested master plus a commit from updateImageryFromJOSM because I have recently put MA-specific imagery in ELI which has flowed to JOSM.  But it's literally just updating the imagery geojson file.

**Comment by simonpoole** [comment]
> 
> 
> For pedantry I should say that I tested master plus a commit from updateImageryFromJOSM because I have recently put MA-specific imagery in ELI which has flowed to JOSM. 

You don't actually have to commit that, as you can simply update from both source in the app.

**Comment by gdt** [comment]
I went out and did ~30s tracklogs and point-from-GPS creation at 3 reference marks, several times.   Things worked pretty well, except that I noticed that after pushing 'pause gpx', I saw a "socket closed" toast.   Sure enough, I had no more position updates.  I think they came back after about a minute, but I often moved to next mark and then enabled recording.  I think they also might have come back from flipping focus to launcher and back.

So my wild guess is that the socket is enabled when wanting position when vespucci gets focus, and is disabled by stopping recording, instead of disabled if "vespucci loses focus AND not recording".

(The current code works better than the old code, and much better than my addition of canceling.  I'm happy to look into this myself, if you want to just not for a while.)

**Comment by simonpoole** [comment]
I was being a bit to aggressive with cancelling the NMEA input (due to the way things work they would get restarted, but cancelling them is unnecessary to start with),  tentatively fixed in https://github.com/MarcusWolschon/osmeditor4android/commit/8ca8152a4d878d594aadd424c249de81d9f2680b

**Comment by gdt** [comment]
I tested briefly and things looked good.    Home button while not recording caused the external device to indicate that the TCP connection was closed, and it stayed on when recording.    I didn't get any errors and I had position when I expected it.   That said, this was a brief indoor test.   I am going to close this and will open  a new issue if I find anything else in further testing/use.    Thanks very much for digging into my problem.
